src/useradd.c

```diff
@@ -2319,6 +2319,7 @@ static void create_home (void)
 					Prog, path);
 				fail_exit(E_HOMEDIR);
 			}
+			free(btrfs_check);
```

Comment:
I would love to see a PR in the future that changes such things (which are not returned and must always be freed) to allocating the way lxc does:

```
/* free allocated memory */
__do_free __attribute__((__cleanup__(__auto_free__)))
{
   __do_free char *btrfs_check = NULL;
}
```

Reply:
Hmmm, didn't know about that attribute.

<https://gcc.gnu.org/onlinedocs/gcc/Common-Variable-Attributes.html#index-cleanup-variable-attribute>

I'll keep your wish in mind for some day.  :)

I've always hated automatic cleanup from C++, because of ETOOMUCHMAGIC, and handling corner cases, but in this case, since you specify it at the top of the function, it's not so magic.

Reply:
you don't have to specify it at the top of the func, only in the decl of the var.  the thing i dislike is that you have to declare a helper API for every type/free combo.  i understand why (lack of generics in C), it's just annoying amount of boilerplate, and easy to misuse if you aren't careful with your type matching.

https://github.com/google/minijail/commit/e933fcee1c215093dd9f349a82993d4b90773a46 is an example of adding a helper for string functions.

Reply:
> you don't have to specify it at the top of the func, only in the decl of the var.

Right.  I try to restrict myself to C89 declarations and short functions, so they usually happen to be at the top of functions.  :)

> the thing i dislike is that you have to declare a helper API for every type/free combo. i understand why (lack of generics in C), it's just annoying amount of boilerplate, and easy to misuse if you aren't careful with your type matching.
> 
> [google/minijail@e933fce](https://github.com/google/minijail/commit/e933fcee1c215093dd9f349a82993d4b90773a46) is an example of adding a helper for string functions.

Thanks!

Reply:
I like it for free() and close(), things which I too often see missed...

Reply:
I was playing with it and can't make it work.  Am I doing anything wrong?

```c
$ cat cleanup.c 
#include <stdlib.h>

int
main(void)
{
	[[gnu::cleanup(free)]] void  *p = malloc(1);
}
```
```c
$ gcc -Wall -Wextra cleanup.c 
cleanup.c: In function ‘main’:
cleanup.c:6:39: warning: ‘free’ called on unallocated object ‘p’ [-Wfree-nonheap-object]
    6 |         [[gnu::cleanup(free)]] void  *p = malloc(1);
      |                                       ^
cleanup.c:6:39: note: declared here
    6 |         [[gnu::cleanup(free)]] void  *p = malloc(1);
      |                                       ^
```

Reply:
Ahh yeah, this things gets a pointer to a pointer.